File: src/org/cerion/symcalc/expression/function/core/N.java
package org.cerion.symcalc.expression.function.core;

import org.cerion.symcalc.expression.Expr;
import org.cerion.symcalc.expression.function.Function;
import org.cerion.symcalc.expression.function.FunctionExpr;
import org.cerion.symcalc.expression.number.IntegerNum;

public class N extends FunctionExpr {

    public N(Expr... e) {
        super(Function.N,e);
    }

    @Override
    protected Expr evaluate() {
        if(size() > 0) {
            if(size() > 1 && get(1).isInteger()) {
                IntegerNum n = (IntegerNum)get(1);
                getEnv().setNumericalEval(true, n.intValue());
            } else
                getEnv().setNumericalEval(true);

            return get(0).eval();
        }

        return this;
    }
}

File: symcalc/src/main/kotlin/function/core/N.kt
package org.cerion.symcalc.function.core

import org.cerion.symcalc.expression.ConstExpr
import org.cerion.symcalc.expression.Expr
import org.cerion.symcalc.expression.VarExpr
import org.cerion.symcalc.function.FunctionExpr
import org.cerion.symcalc.function.FunctionFactory
import org.cerion.symcalc.number.NumberExpr
import org.cerion.symcalc.number.RealBigDec

class N(vararg e: Any) : FunctionExpr(*e) {

    override fun evaluate(): Expr {
        val e = args[0]
        val machinePrecision = size == 1
        val precision = if (machinePrecision) MachinePrecision else get(1).asInteger().intValue()

        if (precision == 0)
            return RealBigDec.ZERO

        if (e is VarExpr)
            return e
        if (e is ConstExpr)
            return e.evaluate(precision)
        if (e is NumberExpr)
            return e.toPrecision(precision)

        if(e is FunctionExpr) {
            val newArgs = mutableListOf<Expr>()
            for (i in 0 until e.args.size) {
                newArgs.add(e.args[i].eval(precision))
            }
            return FunctionFactory.createInstance(e.name, *newArgs.toTypedArray()).eval()
        }

        return this
    }

    override fun validate() {
        validateParameterRange(1, 2)
    }
}
